garglk/sndsdl.c

```diff
@@ -247,54 +263,163 @@ glui32 glk_schannel_play(schanid_t chan, glui32 snd)
+    }
+
+    return successes;
 }
 
 void glk_sound_load_hint(glui32 snd, glui32 flag)
 {
     /* nop */
 }
 
-void glk_schannel_set_volume(schanid_t chan, glui32 vol)
+/** Make an incremental volume change when the fade timer fires */
+Uint32 volume_timer_callback(Uint32 interval, void *param)
 {
+    schanid_t chan = (schanid_t)param;
+
     if (!chan)
     {
-        gli_strict_warning("schannel_set_volume: invalid id.");
-        return;
+        gli_strict_warning("volume_timer_callback: invalid channel.");
+        return 0;
     }
-    chan->volume = vol > 0x10000 ? MIX_MAX_VOLUME : round(pow(((double) vol) / 0x10000, log(4)) * MIX_MAX_VOLUME);
-    switch (chan->status)
+
+    chan->float_volume += chan->volume_delta;
+
+    if (chan->float_volume < 0)
+        chan->float_volume = 0;
+
+    if ((int)chan->float_volume != chan->volume)
     {
-        case CHANNEL_IDLE:
-            break;
-        case CHANNEL_SOUND:
+        chan->volume = (int)chan->float_volume;
+
+        if (chan->status == CHANNEL_SOUND)
             Mix_Volume(chan->sdl_channel, chan->volume);
-            break;
-        case CHANNEL_MUSIC:
+        else if (chan->status == CHANNEL_MUSIC)
             Mix_VolumeMusic(chan->volume);
-            break;
     }
+
+    chan->volume_timeout--;
+
+    /* If the timer has fired FADE_GRANULARITY times, kill it */
+    if (chan->volume_timeout <= 0)
+    {
+        if (chan->volume_notify)
+        {
+            gli_event_store(evtype_VolumeNotify, 0,
+                0, chan->volume_notify);
+        }
+
+        if (!chan->timer)
+        {
+            gli_strict_warning("volume_timer_callback: invalid timer.");
+            return 0;
+        }
+        SDL_RemoveTimer((SDL_TimerID)chan->timer);
+        chan->timer = 0;
+
+        if (chan->volume != chan->target_volume)
+        {
+            chan->volume = chan->target_volume;
+            if (chan->status == CHANNEL_SOUND)
+                Mix_Volume(chan->sdl_channel, chan->volume);
+            else if (chan->status == CHANNEL_MUSIC)
+                Mix_VolumeMusic(chan->volume);
+        }
+        return 0;
+    }
+
+    return interval;
+}
+
+/** Start a fade timer */
+void init_fade(schanid_t chan, int volume, int duration, int notify)
+{
+    if (!chan)
+    {
+        gli_strict_warning("init_fade: invalid channel.");
+        return;
+    }
+
+    chan->volume_notify = notify;
+
+    /* Convert requested Glk Volume to SDL Volume */
+    chan->target_volume = volume >= GLK_MAXVOLUME ? MIX_MAX_VOLUME : round(pow(((double) volume) / GLK_MAXVOLUME, log(4)) * MIX_MAX_VOLUME);
+
+    chan->float_volume = (double)chan->volume;
+    chan->volume_delta = (double)(chan->target_volume - chan->volume) / FADE_GRANULARITY;
+
+    chan->volume_timeout = FADE_GRANULARITY;
+
+    if (chan->timer)
+        SDL_RemoveTimer((SDL_TimerID)chan->timer);
+
+	chan->timer = (int)SDL_AddTimer((Uint32)(duration / FADE_GRANULARITY), volume_timer_callback, (void *)chan);
```

Comment:
This is indented with a tab instead of 4 spaces.

Reply:
In addition, `chan->timer` is an `int` and `SDL_AddTimer` returns a pointer.  On many 64-bit systems this could result in the truncation of the pointer.  `timer` should probably just be made an `SDL_TimerID` instead of `int`.  I think you'd have to add a macro, e.g. `-DGARGLK_USESDL` when building with SDL, so you can conditionally include `SDL/SDL_timer.h` in `garglk/garglk.h` and conditionally define `timer`.  In fact, you could probably conditionally include most/all of `struct glk_schannel_struct` so the members don't exist unless SDL is being used for sound.

Reply:
Thanks for the review! I actually think SDL_TIMERID changes from a pointer to an int between SDL versions, which is why the code ended up this way, but not really an excuse. I'll see if I can work out how to fix it properly.